Make it possible to search multiple markers (Fixes #1738)

brisad opened the pull request.

brisad commented:
Oops! :man_facepalming:   Good catch finding that regression!
It is the global flag `g` in `RegExp` that causes the `.test()` methods to be stateful. I'll fix it and add a test.

Sure, I can add support for network markers too, no problem. Then we can have the same hint in all search fields. :+1: Though if an URL has a comma in it, we'll be unable to search for the actual comma unless we do something like add support for escaping. But perhaps it is nothing to worry about?

julienw commented:
> Though if an URL has a comma in it, we'll be unable to search for the actual comma unless we do something like add support for escaping. But perhaps it is nothing to worry about?

Let's not focus on this edge case, this is too rare to be useful :)

brisad commented:
@gregtatum I've fixed the regression, added comma support for network markers, and added the hint to all search fields.

To do the last point, I changed the `StackSearchField` (which had the hint) to a more general, unconnected component `PanelSearch`. That one could then be reused in all panels. Hope that solution works out.

brisad commented on src/components/shared/MarkerSettings.css:
```diff
@@ -14,20 +14,10 @@
-.markerSettingsSearchbar {
-  padding: 0 5px;
+.markerSettingsSearchField .panelSearchFieldInput {
+  width: 250px;
```
Can this be done better? To shorten the width for the marker settings search field, it reaches into class names that really should be implementation details for the `PanelSearch` component.

gregtatum replied:
Hmm... I'm not sure my favorite pattern for this. I've done it where you can inject a custom class name, or assign that value programmatically with a CSS variable. I think this is fine though. It's a little "ugly" in that we have some nesting, but it's a simple fix.

gregtatum commented on src/components/shared/NetworkSettings.css:
```diff
@@ -14,20 +14,10 @@
 
-.networkSettingsSearchField {
-  width: 250px;
+.networkSettingsSearchField .panelSearchFieldIntroduction {
```
Optional (no need to re-review if you change it)

On my mac the text has an orphan word that goes to the newline. It might be nice to make it all fit on one, although this isn't really guaranteed on other systems.
<img width="351" alt="image" src="https://user-images.githubusercontent.com/1588648/60624997-40e20200-9dac-11e9-9b55-238d29badb42.png">

brisad replied:
It looks the same on my Linux system.
But if I widen the introduction box to make all text fit on one line, the label and the input box above will be displaced so that the input will no longer be completely aligned to the right.

The simple fix is to also widen the input a bit, is it fine if I do that?

julienw replied:
I think it's fine to make the search box wider, like it is in the call tree.

brisad commented on src/profile-logic/marker-data.js:
```diff
@@ -79,51 +79,52 @@ export function deriveJankMarkers(
       if (
         typeof data.category === 'string' &&
-        data.category.toLowerCase().includes(lowerCaseSearchString)
+        searchRegExp.test(data.category)
```
I'm getting a bit curious too, now that you mention it. I think that I, in the back of my head, assumed that a regex would be faster when using or-ed together words; compared to having to loop through them. But I have no idea really.

I might do a measurement later, in that case I'll let you know what I found :)

julienw replied:
In my mind, a regexp is fast if it's reused like it is now; a fixed cost of a regexp is when it's first used and must be compiled. (but I might be wrong :) ).

In the call tree we used both strategies: look for `filterThreadToSearchStrings` in `profile-data.js` where we loop through all search strings to filter the call tree, but we used the regexp to highlight in the TreeView component. I don't know if that's the best strategy, maybe using the regexp always would have been just as good, but we don't know unless we profile with a bit profile.

Some heavy profiles to make profiling easier:
[profile with a lot of network markers](https://profiler.firefox.com/public/52d5d452191340ab79ac92b519bcac4479fc3ab9/network-chart/?globalTrackOrder=0-1-2-3&thread=4&v=3)
[profile with an awful lot of markers](https://profiler.firefox.com/public/1c77f550d099ab67995681cf835ae589dd88e714/marker-chart/?globalTrackOrder=0-1-2-3&thread=4&v=3)

brisad replied:
@gregtatum @julienw I did a very quick and dirty comparison on the profile with an awful lot of markers. I compared the current RegExp-code with a version where I instead looped over an array of search words where I repeatedly called `toLowerCase()` and used `include()` to find matches.

I tried searching with the comma separated example search term `"User,GC,Idle"` and measured the time spent in `getSearchFilteredMarkerIndexes`.

It took 5 ms with the RegExp and 95 ms without. Much of the time in the non-RegExp version is spent in `std::vector::_M_realloc_insert`, so I'm guessing that we get a lot of allocations when calling `toLowerCase()` all the time.

Even when I removed all calls to `toLowerCase()` it still took 25 ms. Seems like using a RegExp here is pretty efficient.

brisad commented:
Thanks for the review! Just one question before merging.
